## EJERCICIOS DEL CUESTIONARIO

## pregunta 5 - cuestionario

In [29]:
from sklearn import preprocessing, model_selection
from sklearn.neural_network import MLPClassifier

In [30]:
# configuración para notebook con instalación LOCAL
FUENTES_DIR = '../Fuentes' # carpeta donde se encuentran archivos .py auxiliares
DATOS_DIR   = '../Datos/'  # carpeta donde se encuentran los datasets

# agrega ruta de busqueda donde tenemos archivos .py
import sys
sys.path.append(FUENTES_DIR)

In [6]:
import pandas as pd      # para trabajar con archivos de datos csv, excel, etc: https://pandas.pydata.org/docs/getting_started/tutorials.html
#import chardet           # para detectar la codificación de texto en archivos
import numpy as np


nombre_archivo = DATOS_DIR + 'Vinos.csv' 

# recupera el archivo en un objeto dataframe de pandas utilizando la codificación detectada
df = pd.read_csv(nombre_archivo, sep=';')
print(df)

     Class  Alcohol  Malic acid   Ash  Alcalinity of ash  Magnesium  \
0        1    14.23        1.71  2.43               15.6        127   
1        1    13.20        1.78  2.14               11.2        100   
2        1    13.16        2.36  2.67               18.6        101   
3        1    14.37        1.95  2.50               16.8        113   
4        1    13.24        2.59  2.87               21.0        118   
..     ...      ...         ...   ...                ...        ...   
173      3    13.71        5.65  2.45               20.5         95   
174      3    13.40        3.91  2.48               23.0        102   
175      3    13.27        4.28  2.26               20.0        120   
176      3    13.17        2.59  2.37               20.0        120   
177      3    14.13        4.10  2.74               24.5         96   

     Total phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
0             2.80        3.06                  0.28             2.29   
1

In [6]:
# %% separa atributos y clases
X_raw = np.array(df.iloc[:,1:])  # recupera todas las columnas salvo la primera (es la clase)
Y_raw = np.array(df.iloc[:,0])    # recupera solo la primer columna (es la clase)

# Binarizador para convertir el nombre de la clase en one hot encoding
binarizer = preprocessing.LabelBinarizer()

# Binariza cada clase como una combinación de un 1 y 0s
Y_raw = binarizer.fit_transform(Y_raw)
# Y_raw==pd.get_dummies(df[' Balance']).to_numpy() # forma alternativa para codificar

print('Las clases del dataset son :', binarizer.classes_)
print(Y_raw)

########### APLICO NORMALIZACION CON MEDIA Y DESVIO ################3
scaler = preprocessing.MinMaxScaler()
X_raw  = scaler.fit_transform( X_raw )

Las clases del dataset son : [1 2 3]
[[1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]


In [16]:
# f.activacion --> 'relu
#alfa = 0.1
#CotaError = 10e-5
#MAX_ITE = 30

modelo= MLPClassifier(solver='sgd', learning_rate_init=0.1,

                 hidden_layer_sizes=(6, ), random_state=1,

                 max_iter=30, verbose=False,  tol=10e-05,

                 activation='relu')

modelo.fit(X_raw, Y_raw)


#  ########### Medición del entrenamiento ######################
# entrenamiento con X_train e Y_train 
Y_pred = modelo.predict(X_raw)
score = modelo.score(X_raw, Y_raw)

# "invierte" la transformacion binaria para obtener los nombres de las clases
Y_it = binarizer.inverse_transform(Y_raw)
Y_pred_it = binarizer.inverse_transform(Y_pred)

# calculo manual del accuracy
print('Efectividad: %6.2f%%' % (100*(Y_pred_it == Y_it).sum()/len(Y_it)) )
print('      Score: %6.2f%%' % (score) )


Efectividad:  93.82%
      Score:   0.92%


C:\Users\W10\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (30) reached and the optimization hasn't converged yet.
  warnings.warn(


### el accuracy es del 93.82% , no del 100%

### --------------------------------------------------------------------------

### Preguntas 7 y 8 - cuestionario

In [40]:
import tensorflow as tf
from tensorflow.keras.models import load_model
from tensorflow.keras.layers import LeakyReLU

# Cargar el modelo desde el archivo
model = load_model(DATOS_DIR + "mod_HOUSING.h5", custom_objects={'LeakyReLU': LeakyReLU, 'loss': 'mean_squared_error'})

# Mostrar el resumen del modelo
model.summary()


ValueError: Unrecognized keyword arguments: ['batch_shape']

### Si se aplica el modelo a todos los ejemplos de entrenamiento se obtiene un valor del coeficiente de determinación R2 superior a 0.9

In [36]:
# importo csv de datos housing.csv
nombre_archivo = DATOS_DIR + 'housing.csv' 

# recupera el archivo en un objeto dataframe de pandas utilizando la codificación detectada
df = pd.read_csv(nombre_archivo, sep=',')
print(df)

       longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  media

In [37]:
#elimino columna "ocean_proximity" porque no se usa:
df.drop("ocean_proximity", axis=1, inplace=True)

# cuento cuantos valores faltantes:
df.isnull().sum()


       longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  media

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
dtype: int64

In [32]:
#elimino filas c valores faltantes
df.dropna(inplace=True)
#verifico:
df.isnull().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
dtype: int64

In [38]:
# %% separa atributos y clases
X = np.array(df.iloc[:,:-1])  # recupera todas las columnas salvo la ultima
Y = np.array(df.iloc[:,-1])    # recupera solo la ultima columna (valor a predecir)

#normalizo con StandardScaler (usando media y desvio)
# Escala los valores entre 0 y 1
min_max_scaler = preprocessing.StandardScaler()
X_train = min_max_scaler.fit_transform(X)


In [39]:
from sklearn.metrics import r2_score

# Supongamos que Y_test son tus valores reales y Y_pred son predicciones del modelo
Y_pred = model.predict(X_train)  # Predicciones del modelo

# Calcular R^2
r2 = r2_score(Y, Y_pred)

print(f"El valor de R^2 es: {r2}")

# Comprobar si es mayor a 0.9
if r2 > 0.9:
    print("El modelo tiene un buen ajuste (R^2 > 0.9)")
else:
    print("El modelo no tiene un buen ajuste (R^2 <= 0.9)")


645/645 [==============================] - 1s 1ms/step


ValueError: Input contains NaN, infinity or a value too large for dtype('float32').